Docs: Add Writing Stories in TypeScript docs guide

kylegach opened the pull request:
Closes #22595, closes #23362

<!-- Thank you for contributing to Storybook! If your PR is related to an issue, provide the number(s) above; if it resolves multiple issues, be sure to break them up (e.g. "closes #1000, closes #1001"). -->

## What I did

- Add _Writing Stories in TypeScript_ (`/writing-stories/typescript`) docs guide
- Use same snippets for all "custom args" examples
- Add/update/remove snippets

## How to test

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/react/contribute/new-snippets#preview-your-work) for this branch, `writing-stories-in-typescript`

## Checklist

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to make sure your PR is ready to be reviewed. -->

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [x] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "build", "documentation", "maintenance", "dependencies", "other"]`

<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/react/contribute/how-to-contribute

-->

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

kylegach committed:
Add Writing Stories in TypeScript docs guide

- Use same snippets for all "custom args" examples
- Add/update/remove snippets

kasperpeulen commented on docs/snippets/angular/page-story-slots.ts.mdx:
```diff
@@ -5,26 +5,27 @@ import type { Meta, StoryObj } from '@storybook/angular';
-
-export default meta;
-type Story = StoryObj<Page>;
+// TODO: Is this correct?
+//       If not, is it expected that an Angular component exports its props type?
+//       What if it's an interface instead of a type (which seems to be the convention)?
+//       Or, how do you extract an Angular component's props type?
+type PagePropsAndCustomArgs = Page & { footer?: string };
```
What do you mean with this interface convention? I haven't heard of that.

I think the component will always contain all props (and more), so it kind of works, but users might get autocompletion for args that don't really exist.

I guess that is an acceptable tradeoff? For people that want really strict types, they would need to do something like:
`type Story = Pick<Page, 'prop1' | 'prop2'>`

I don't think the utility you want exists, I also asked it here:
https://github.com/storybookjs/storybook/discussions/20988

kasperpeulen commented on docs/snippets/solid/page-story-slots.ts.mdx:
```diff
@@ -5,24 +5,22 @@ import type { Meta, StoryObj } from 'storybook-solidjs';
 import { Page } from './Page';
 
-const meta: Meta<typeof Page> = {
+// TODO: How do you extract a Solid component's props type?
+// type PagePropsAndCustomArgs = Page & { footer?: string };
```
Good question, I don't know!

kasperpeulen commented on docs/snippets/vue/page-story-slots.3.ts-4-9.mdx:
```diff
@@ -5,19 +5,11 @@ import type { Meta, StoryObj } from '@storybook/vue3';
 
 import Page from './Page.vue';
 
+// TODO: How do you extract a Vue3 component's props type?
+// type PagePropsAndCustomArgs = Page & { footer?: string };
```
We use 
https://github.com/vuejs/language-tools/tree/master/packages/vue-component-type-helpers

Not sure if we should add that to the docs though.. Hmm 🤔

kasperpeulen commented on docs/snippets/web-components/page-story-slots.ts.mdx:
```diff
@@ -5,20 +5,23 @@ import type { Meta, StoryObj } from '@storybook/web-components';
 import { html } from 'lit';
 
-const meta: Meta = {
+// TODO: How do you extract a web component's props type?
+// type PagePropsAndCustomArgs = Page & { footer?: string };
```
Same story as angular, I think this is probably fine.

kasperpeulen commented on docs/snippets/vue/page-story-slots.3.ts.mdx:
```diff
@@ -5,19 +5,11 @@ import type { Meta, StoryObj } from '@storybook/vue3';
-
-export default meta;
-type Story = StoryObj<typeof Page>;
+// TODO: How do you extract a Vue3 component's props type?
+// type PagePropsAndCustomArgs = Page & { footer?: string };
```
I don't think there exist something official, I do some conditional typing magic in the implementation:
```ts
type ComponentProps<C> = C extends ExtendedVue<any, any, any, any, infer P>
  ? P
  : C extends Component<any, any, any, infer P>
  ? P
  : unknown;
```

kylegach committed:
Address TODOs

jonniebigodes commented on docs/writing-stories/typescript.md:
```diff
@@ -0,0 +1,149 @@
+
+## Typing stories with `Meta` and `StoryObj`
+
+When writing stories, there are two aspects that are helpful to type. The first is the [component meta](./introduction.md#default-export), which describes and configures the component and its stories. In a [CSF file](..api.md/csf), this is the default export. The second is the [stories themselves](./introduction.md#defining-stories).
```
@kylegach when you have a moment, the CSF link is incorrect. Mind fixing it?

jonniebigodes commented on docs/writing-stories/typescript.md:
```diff
@@ -0,0 +1,149 @@
+
+<IfRenderer renderer={['angular', 'web-components']}>
+
+<div class="aside">
+
+ℹ️ We are not yet able to provide additional type safety using the `satisfies` operator with Angular and Web components.
+
+<details>
+<summary>More info</summary>
+
+Both Angular and Web components utilize a class plus decorator approach. The decorators provide runtime metadata, but do not offer metadata at compile time.
+
+As a result, it appears impossible to determine if a property in the class is a required property or an optional property (but non-nullable due to a default value) or a non-nullable internal state variable.
+
+For more information, please refer to [this discussion](github.com/storybookjs/storybook/discussions/20988).
+
+</details>
```
@kylegach while previewing this locally, it reads a bit weird and adds a bit of confusion with the amount of information you're trying to present. A couple of options we could here:
1- Condense the information in a single paragraph without the aside and provide the link for the relevant discussion (don't forget to update the link as it throws a 404).
2 - Push this aside into a troubleshooting section visible only for the frameworks (i.e., Angular and Web Components) with some word adjustments

kylegach replied:
I fixed the link (thanks for the catch!). But I still like the current approach better than either of those options.

 - The aside makes it immediately clear to the affected users that `satisfies` is not available.
 - Rather than just taking our word for it, the "More info" provides the _why_.
     - While this could be an anchor link to a conditionally rendered troubleshooting section, that seemed far more clunky than a details disclosure.
 - I don't think we should combine the initial message with what's in "More info", because _that_ is too much information to present upfront.

jonniebigodes commented on docs/writing-stories/typescript.md:
```diff
@@ -0,0 +1,149 @@
+
+## Framework specific tips
+
+Template-based frameworks such as Vue and Svelte typically require editor extensions to enable syntax highlighting, autocomplete, and type checking. Here are a few tips to help you set up the ideal environment for them.
```
Mind taking a second pass at this and restructuring the documentation a bit, as this is going to be applied only for Vue and Svelte; it seems that the heading seems a bit redundant when a user is already previewing the documentation for any of those frameworks.

kylegach replied:
Thanks for the push toward simplicity!

jonniebigodes commented on docs/writing-stories/typescript.md:
```diff
@@ -0,0 +1,149 @@
+
+<!-- prettier-ignore-end -->
+
+### Optional type parameter
```
Can you elaborate on the need for this heading when the text within it relates to the above? Couldn't it be removed and the text below adjusted to provide the information explicitly? More even as we're teaching users to use this pattern in all the examples we provide, and adding this may add some confusion simply by referencing it as optional. 

What do you think?

jonniebigodes commented on docs/writing-stories/typescript.md:
```diff
@@ -0,0 +1,149 @@
+title: 'Writing stories in TypeScript'
+---
+
+Writing your stories in [TypeScript](https://www.typescriptlang.org/) makes you more productive. You don't have to jump between files to look up component props. Your code editor will alert you about missing required props and even autocomplete prop values. Plus, Storybook infers those component types to auto-generate the [Controls](../api/doc-block-controls.md) table.
```
Could you take another pass at the introduction and condense it a bit to be more straight to the point, as the way it's written, it seems that we're repeating the same idea.

jonniebigodes reviewed (changes requested):
Appreciate you taking the time to put together this great addition to the documentation, and can't wait to get it out and made available to our readers. Left a couple of items for you to look at when you have a chance.

Let me know and we'll go from there.

kylegach committed:
Address comments

- Simplify introduction
- Fix links
- Tweak "type parameter" heading & text
- Fix typos
- Simplify renderer-specific content structure
